### PR TITLE
Private Channels

### PR DESCRIPTION
**Pull request by niftynei**
Allows users to create private channels, solves #2125

### CODE REVIEW

File: lightningd/opening_control.c [hunk 1]
```diff
@@ -780,6 +781,7 @@ static void json_fund_channel(struct command *cmd,
 		   p_req("id", json_tok_pubkey, &id),
 		   p_req("satoshi", json_tok_tok, &sattok),
 		   p_opt("feerate", json_tok_feerate, &feerate_per_kw),
+		   p_opt_def("private", json_tok_bool, &private_channel, false),
```
Comment (rustyrussell) [inline lightningd/opening_control.c]: The spec calls this bit `announce_channel`; should we invert this and call it "announce", default=true?

File: lightningd/peer_control.c [hunk 2]
```diff
@@ -542,21 +542,6 @@ static enum watch_result funding_lockin_cb(struct lightningd *ld,
...
-	 *     - MUST send the `announcement_signatures` message.
-	 *       - MUST NOT send `announcement_signatures` messages until
-	 *         `funding_locked` has been sent AND the funding transaction has
-	 *         at least six confirmations.
-	 *   - otherwise:
-	 *     - MUST NOT send the `announcement_signatures` message.
-	 */
-	if (!(channel->channel_flags & CHANNEL_FLAGS_ANNOUNCE_CHANNEL))
-		return DELETE_WATCH;
-
```
Comment (rustyrussell) [inline lightningd/peer_control.c]: This change is probably OK, but I'd like something in the commit msg justifying its deletion; it seems orthogonal?

File: tests/test_connection.py [hunk 3]
```diff
@@ -791,6 +791,21 @@ def test_channel_persistence(node_factory, bitcoind, executor):
     l1.daemon.wait_for_log(' to ONCHAIN')
 
 
+def test_private_channel(node_factory):
+    l1, l2 = node_factory.line_graph(2, fundchannel=True, announce=True, privatechannels=True)
+    l3, l4 = node_factory.line_graph(2, fundchannel=True, announce=True, privatechannels=False)
```
Comment (rustyrussell) [inline tests/test_connection.py]: fundchannel defaults True already AFAICT?  We should probably rename the 'announce' parameter to 'wait_for_announce'?

File: tests/utils.py [hunk 4]
```diff
@@ -839,9 +839,10 @@ def both_dirs_ready(n, scid):
         wait_for(lambda: both_dirs_ready(nodes[-1], scids[0]))
 
         # Make sure we have all node announcements, too (just check ends)
-        for n in nodes:
-            for end in (nodes[0], nodes[-1]):
-                wait_for(lambda: 'alias' in only_one(end.rpc.listnodes(n.info['id'])['nodes']))
+        if not privatechannels:
```
Comment (rustyrussell) [inline tests/utils.py]: Hmm announce=True and privatechannels=True is an invalid combination (which kinda reinforces my preference to call them 'wait_for_announce' and 'announce' BTW)  Just assert() above that they're not asking for it?

File: doc/lightning-fundchannel.7.txt [hunk 5]
```diff
@@ -31,6 +31,9 @@ than 16777215 satoshi.
 'urgent', 'normal' or 'slow' to use lightningd's internal estimates:
 'normal' is the default.
 
+'private' is an optional flag to keep this channel private, i.e. not
+broadcast a `channel_announcement` message.
```
Comment (rustyrussell) [inline doc/lightning-fundchannel.7.txt]: s/don't offer to create a `channel_announcement` message.  Note that we won't create a channel_announcement message unless the peer agrees to as well./?

File: CHANGELOG.md [hunk 6]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - JSON API: `getinfo` now returns `num_peers` `num_pending_channels`,
   `num_active_channels` and `num_inactive_channels` fields.
 - JSON API: use `\n\n` to terminate responses, for simplified parsing (pylightning now relies on this)
+- JSON API: `fundchannel` now includes option to keep channel private
```
Comment (rustyrussell) [inline CHANGELOG.md]: Name of option, and note default is public?

File: lightningd/opening_control.c (hunk 1 above)
Comment (niftynei) [reply lightningd/opening_control.c]: works for me. :+1:

File: lightningd/peer_control.c (hunk 2 above)
Comment (niftynei) [reply lightningd/peer_control.c]: reverted. I was under the mistaken assumption that this would prevent private channels from being locked. Turns out that check is the line above...

File: tests/utils.py (hunk 4 above)
Comment (niftynei) [reply tests/utils.py]: Great point. Updated.

File: CHANGELOG.md (hunk 6 above)
Comment (niftynei) [reply CHANGELOG.md]: Updated.

File: doc/lightning-fundchannel.7.txt (hunk 5 above)
Comment (rustyrussell) [reply doc/lightning-fundchannel.7.txt]: Damn, you're right!  Oops, I don't know how I got that idea then.  Thanks!

File: tests/utils.py [hunk 7]
```diff
@@ -839,9 +840,10 @@ def both_dirs_ready(n, scid):
         wait_for(lambda: both_dirs_ready(nodes[-1], scids[0]))
 
         # Make sure we have all node announcements, too (just check ends)
-        for n in nodes:
-            for end in (nodes[0], nodes[-1]):
-                wait_for(lambda: 'alias' in only_one(end.rpc.listnodes(n.info['id'])['nodes']))
+        if not announce_channels:
+            for n in nodes:
+                for end in (nodes[0], nodes[-1]):
+                    wait_for(lambda: 'alias' in only_one(end.rpc.listnodes(n.info['id'])['nodes']))
```
Comment (rustyrussell) [inline tests/utils.py]: Can't happen?  We have `if not wait_for_announce: return` above

Comment (niftynei) [reply tests/utils.py]: ack.

**Comment by niftynei** [comment]
self ACK https://github.com/ElementsProject/lightning/pull/2147/commits/8a1fc7b834a0a222921ed9197e27dd2b1fb25164

**Comment by niftynei** [comment]
ACK https://github.com/ElementsProject/lightning/pull/2147/commits/7129e99584e76e40ac271a1a690d3ba29f122f6c

**Comment by shesek** [comment]
Is there a way to tell if an existing channel is private or not? I can't seem to find anything indicating this in `listpeers`.

**Comment by niftynei** [comment]
`listchannels` has a `public` boolean which should tell you if a channel is private or not. it'd make sense to add this to the channel obj in `listpeers` as well.
